test/ejabberd_c2s_SUITE.erl

```diff
@@ -88,6 +89,57 @@ send_error_when_waiting(_) ->
     ?eq(close, Close),
     ok.
 
+c2s_is_killed_when_to_many_messages_in_the_queue(_) ->
+    meck:new(ejabberd_c2s, [passthrough]),
+    MaxQueueSize = 50,
+    Self = self(),
+    %% We similate a long running event during which
+    %% there will be many messages put into C2S process message queue
+    meck:expect(ejabberd_c2s, handle_event,
+                fun(go_to_sleep, StateName, ProcState) ->
+                        Self ! c2s_going_to_sleep,
+                        ct:pal("going to sleep"),
+                        timer:sleep(2000),
+                        {next_state, StateName, ProcState};
+                   (Event, StateName, ProcState) ->
+                        meck:passthrough([Event, StateName, ProcState])
+                end),
+    {ok, C2SPid} = given_c2s_started([{max_fsm_queue, MaxQueueSize}]),
+
+    %% We want to monitor the c2s process and not being linked to it
+    Ref = erlang:monitor(process, C2SPid),
+    erlang:unlink(C2SPid),
+
+    p1_fsm_old:send_all_state_event(C2SPid, go_to_sleep),
+
+    receive
+        c2s_going_to_sleep ->
+            ok
+    end,
+
+    meck:unload(ejabberd_c2s),
+
+    %% We put MaxQueueSize + 1 messages to C2S Process message queue
+    %% while it is asleep
+    %% The process will be killed when it wakes up and tries to process
+    %% next message
+
+    [p1_fsm_old:send_all_state_event(C2SPid, {event, I}) ||
+     I <- lists:seq(1, 50 + 1)],
```

Comment:
We could use `MaxQueueSize + 1` here to avoid a discrepancy if the value is changed from 50 to something else.